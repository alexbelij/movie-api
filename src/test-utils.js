/* @flow */

import R from 'ramda';

const childTypesForArray = R.memoize(
  R.pipe(R.map(R.type), R.uniq),
);

const modelFromObject = (obj: ?Object) => R.map(
  R.cond([
    [
      R.allPass([
        R.pipe(R.type, R.equals('Array')),
        R.pipe(childTypesForArray, R.length, R.equals(1)),
      ]),
      (val: Array<any>) => `Array<${R.pipe(childTypesForArray, R.head)(val)}>`,
    ],
    [R.pipe(R.type, R.contains(R.__, ['Object', 'Array'])), modelFromObject],
    [R.T, R.type],
  ]),
  obj,
);

const modelFromFirstElement = R.pipe(
  R.head,
  modelFromObject,
);
const modelFromFirstElementAtPath = (path: string) => R.pipe(
  R.pathOr([], path.split('.')),
  modelFromFirstElement,
);

export { modelFromObject, modelFromFirstElement, modelFromFirstElementAtPath };
