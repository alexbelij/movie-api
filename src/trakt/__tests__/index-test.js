/* @flow */

import { modelFromObject } from '../../test-utils';
import Trakt from '../index';

const imdbId = 'tt2488496'; // Star Wars: The Force Awakens
const traktId = 'star-wars-the-force-awakens-2015';

describe('Trakt', () => {
  let trakt: Trakt;

  beforeAll(() => {
    trakt = new Trakt();
  });

  it('fetches movie id by imdb id', async () => {
    expect(await trakt.getId({ imdbId })).toMatchSnapshot();
  });

  it('fetches movie stats for a given id', async () => {
    expect(
      modelFromObject(await trakt.getMovieStats(traktId)),
    ).toMatchSnapshot();
  });

  it('formats movie stats response according to graphql query', async () => {
    const res = await trakt.getMovieStats(traktId, `
      {
        watchers
        votes
      }
    `);

    expect(modelFromObject(res)).toMatchSnapshot();
  });
});
