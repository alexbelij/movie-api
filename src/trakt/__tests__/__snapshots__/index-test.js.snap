exports[`Trakt fetches movie id by imdb id 1`] = `"star-wars-the-force-awakens-2015"`;

exports[`Trakt fetches movie stats for a given id 1`] = `
Object {
  "collectors": "number",
  "comments": "number",
  "lists": "number",
  "plays": "number",
  "votes": "number",
  "watchers": "number",
}
`;

exports[`Trakt formats movie stats response according to graphql query 1`] = `
Object {
  "votes": "number",
  "watchers": "number",
}
`;
